core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -603,5 +640,76 @@ State committed(Substream winningSubstream) {
 
     // GuardedBy RetriableStream.lock
     boolean closed;
+
+    long bufferNeeded;
+    // setting to true must be GuardedBy RetriableStream.lock
+    boolean bufferLimitExceeded;
+  }
+
+
+  /**
+   * Traces the buffer used by a substream.
+   */
+  class BufferSizeTracer extends ClientStreamTracer {
+    // Each buffer size tracer is dedicated to one specific substream.
+    private final Substream substream;
+
+    BufferSizeTracer(Substream substream) {
+      this.substream = substream;
+    }
+
+    /**
+     * A message is sent to the wire, so its reference would be released if no retry or
+     * hedging were involved. So at this point we have to hold the reference of the message longer
+     * for retry, and we need to increment {@code substream.bufferNeeded}.
+     */
+    @Override
+    public void outboundWireSize(long bytes) {
+      if (state.winningSubstream != null || substream.closed) {
+        return;
+      }
+
+      Runnable postCommitTask = null;
+
+      // TODO(zdapeng): avoid using the same lock for both in-bound and out-bound.
+      synchronized (lock) {
+        if (state.winningSubstream != null || substream.closed) {
+          return;
+        }
+        substream.bufferNeeded += bytes;
+        if (substream.bufferNeeded <= perRpcBufferUsed) {
+          return;
+        }
+
+        if (substream.bufferNeeded > perRpcBufferLimit) {
+          substream.bufferLimitExceeded = true;
+        } else {
+          // Only update channelBufferUsed when perRpcBufferUsed is not exceeding perRpcBufferLimit.
+          long savedChannelBufferUsed =
+              channelBufferUsed.addAndGet(substream.bufferNeeded - perRpcBufferUsed);
+
+          if (savedChannelBufferUsed > channelBufferLimit) {
+            substream.bufferLimitExceeded = true;
+          }
+        }
+        perRpcBufferUsed = substream.bufferNeeded;
```

Comment:
This looks wrong for the `substream.bufferNeeded > perRpcBufferLimit` case, because `perRpcBufferUsed` is being updated without also updating `channelBufferUsed`. In `commit()`, the `channelBufferUsed.addAndGet(-perRpcBufferUsed);` will over-adjust the channel buffer usage.

I'd suggest moving this line into the `else`, preferable immediately next to updating `channelBufferUsed`.

Reply:
Corrected with tests added.